scanner/matcher/updater/vuln/updater.go

```diff
@@ -326,17 +335,131 @@ func (u *Updater) runUpdate(ctx context.Context) error {
 		return err
 	}
 
 	if u.initialized.CompareAndSwap(false, true) {
 		zlog.Info(ctx).Msg("finished initial updater run: setting updater to initialized")
 	}
 
+	return nil
+}
+
+// runSingleBundleUpdate updates the vulnerability data.
+func (u *Updater) runMultiBundleUpdate(ctx context.Context) error {
+	ctx = zlog.ContextWithValues(ctx, "component", "matcher/updater/vuln/Updater.runMultiBundleUpdate")
+
+	prevTime, err := u.metadataStore.GetLastVulnerabilityUpdate(ctx)
+	if err != nil {
+		zlog.Debug(ctx).
+			Err(err).
+			Msg("did not get previous vuln update timestamp")
+		return err
+	}
 	zlog.Info(ctx).
-		Str("timestamp", timestamp.Format(http.TimeFormat)).
-		Msg("new vuln update")
+		Time("timestamp", prevTime).
+		Msg("previous vuln update")
+
+	zipFile, zipTime, err := u.fetch(ctx, prevTime)
+	if err != nil {
+		return err
+	}
+	if zipFile == nil {
+		// Nothing to update at this time.
+		zlog.Info(ctx).Msg("no new vulnerability update")
+		return nil
+	}
+	defer func() {
+		if err := zipFile.Close(); err != nil {
+			zlog.Error(ctx).Err(err).Msg("closing temp update file")
+		}
+		if err := os.Remove(zipFile.Name()); err != nil {
+			zlog.Error(ctx).Err(err).Msgf("removing temp update file %q", zipFile.Name())
+		}
+	}()
+
+	// Open the .zip file
+	zipInfo, err := zipFile.Stat()
+	if err != nil {
+		return err
+	}
+	zipReader, err := zip.NewReader(zipFile, zipInfo.Size())
+	if err != nil {
+		return err
+	}
+
+	// Iterate through each vulnerability bundle in the .zip archive
+	names := make([]string, 0, len(zipReader.File))
+	for i := range zipReader.File {
+		bundleF := zipReader.File[i]
+		names = append(names, bundleF.Name)
+		ctx := zlog.ContextWithValues(ctx, "bundle", bundleF.Name)
+		zlog.Info(ctx).Msg("starting bundle update")
+		if err := u.updateBundle(ctx, bundleF, zipTime, prevTime); err != nil {
+			zlog.Error(ctx).Err(err).Msg("updating bundle failed")
+			return fmt.Errorf("updating bundle %s: %w", bundleF.Name, err)
+		}
+		zlog.Info(ctx).Msg("completed bundle update")
+	}
+
+	// Clean updaters that were deleted (not in the zip and older than this update).
+	// Safe to be run concurrently.
+	err = u.metadataStore.GCVulnerabilityUpdates(ctx, names, zipTime)
+	if err != nil {
+		return fmt.Errorf("cleaning vuln updates: %w", err)
+	}
+
+	_ = u.Initialized(ctx)
+
+	return nil
+}
+
+func (u *Updater) updateBundle(ctx context.Context, zipF *zip.File, zipTime time.Time, prevTime time.Time) error {
+	// Use TryLock to prevent simultaneous updates for the same bundle.
+	lCtx, lDone := u.locker.TryLock(ctx, zipF.Name)
+	defer lDone()
+	if err := lCtx.Err(); err != nil {
+		zlog.Info(ctx).Err(err).Msg("skipping: did not obtain lock")
+		return nil
+	}
+
+	// Ensure there is an update timestamp for this bundle, and check if it's newer
+	// then this update.
+	lastTime, err := u.metadataStore.GetOrSetLastVulnerabilityUpdate(lCtx, zipF.Name, prevTime)
```

Comment:
How come we set the time even if there is a potential failure?

Reply:
Setting the timestamp is OK even if this bundle cannot be updated. This means we are holding the last update timestamp to this unless it eventually becomes true. If it fails on this Matcher updater run, fine. We will wait until the next Matcher updater run to move it forward.

Reply:
Sorry I don't quite follow. Can you give an example?